src/python/WMCore/MicroService/Unified/MSMonitor.py

```diff
@@ -0,0 +1,196 @@
+                # move the request status staging -> staged
+                elif self.pileupTransfersCompleted(reqName):
+                    self.logger.debug(
+                        "+++ request %s pileup transfers are completed", req)
+                    self.change(req, 'staged', '+++ monit')
+                # transfers not completed
+                # just update the database with their completion
+                else:
+                    self.logger.debug(
+                        "+++ request %s transfers are %s completed",
+                        req, completed)
+                    # for us workflow name is the same as request name
+                    wname = reqName
+                    requestStatuses[wname] = statusRecords
+            self.updateTransferInfo(requestStatuses)
+        except Exception as err:  # general error
+            self.logger.exception('+++ monit error: %s', str(err))
+
+    def getStatusRecords(self, reqName):
+        """
+        Get transfer statuses for given request name
+        transfer docs on backend has the following form
+        https://gist.github.com/amaltaro/72599f995b37a6e33566f3c749143154
+        So far logic is based on option D of the records
+
+        .. doctest::
+
+            {"workflowName": "bla",
+             "timestamp": 123,
+             "transfers": [rec1, rec2, ... ]
+            }
+            where each record has the following format:
+            {"dataset":"/a/b/c", "dataType": "primary", "transferIDs": [1,2]}
+
+        :param reqName: name of the request
+        :return: list of of status records
+        """
+        statusRecords = []
+
+        # in our case workflow name is the same as request name
+        wname = reqName
+        # get existing transfer IDs record
+        doc = self.getTransferInfo(wname)
+        # loop over all records and obtain new transfer IDs
+        for rec in doc['transfers']:
+            dataset = rec['dataset']
+            dtype = rec['dataType']
+            # obtain new transfer ids for given request name and dataset
+            tids = self.getTransferIds(dataset)
+            rec = {'dataset': dataset, 'dataType': dtype, 'transferIDs': tids}
+            statusRecords += rec
+        return statusRecords
+
+    def pileupTransfersCompleted(self, reqName):
+        "Check if pileup transfers are completed for given request"
+        # in our case workflow name is the same as request name
+        wname = reqName
+        doc = self.getTransferInfo(wname)
+        # loop over all records and obtain new transfer IDs
+        for rec in doc['transfers']:
+            dataset = rec['dataset']
+            ctype = rec['dataType']
+            if ctype == 'pileup':
+                tids = self.getTransferIds(dataset)
+                # if we have any transfer IDs then we still have transfers
+                if tids:
+                    return False
+        return True
+
+    def checkStatusRecords(self, records):
```

Comment:
I'm still not sure I get the usefulness of this method.

Reply:
read the code. It does the following:
- read given records and extract dataset
- it gets new transfer IDs (transfer statuses) for a given dataset
- it save them as new record
- it calculates completion value
- it returns new transferIds (transferStatus) and completion value

Reply:
Which is wrong, see the MS Monitor wiki I provided before (point 4).
BTW, this method is apparently doing the same thing as `getStatusRecords`.

Reply:
And, I insist that we need to get complete status value somehow. I don't mind to remove this method, but you should tell me how to get complete value then?

It is not about persisting records in CouchDB (which is what point 4 is saying) it is about how to find if we completed all transfers. For that we need to take records we put into CouchDB and check them again somehow. The former task (how to get records from Couchdb) is done in `getStatusRecords` while the later task (how to obtain complete status) is done in `checkStatusRecords`. While we obtain a new transfer completion values we can get new transfer IDs (request IDs). That is why checkStatusRecrods is doing both tasks.

Reply:
Let me rephrase it. Monitor wiki says:
```
using the transfer ids available in the transfer document, make calls to PhEDEx/Rucio in order to get the status of those transfers; then calculate the transfer status completion. 
```
The `checkStatusRecords` is doing exactly this:
- take status records (transfer documents)
- it calls `getTransferIds` APi for each record which by itself
- calls PhEDEx subscriptions API
- find out (calculate) completion value

I dont' understand do you want to get rid of this method, rename it or change its logic.

Reply:
I'd rather not mix status records with transfer records. The way I see it is:
* getTransferIds: retrieve the transfer IDs for a given workflow
* checkStatusRecords (or renamed to checkTransferCompletion): uses the transfer IDs, query PhEDEx and calculate the dataset/transfer completion (can be done in multiple methods if you like).

So I see no use for `getStatusRecords`. In addition to that, `checkStatusRecords` is not supposed to generate any new transfer ID, it's supposed to receive transfer IDs, and treat them.

Reply:
ok, I did the following:
- renamed `checkStatusRecords` to `completion` method, so far it does very simple operation. I decided to have it if we need to implement more complex algorithm.
- new method takes transfer records and return completion value